### PR TITLE
MAINT: Tidy PyArray_DescrConverter

### PR DESCRIPTION
**Pull request by eric-wieser**
First part of some major cleanup to make `PyArray_DescrConverter` manageable.

Likely easiest to review commit-by-commit.

---

Follow-up will be to replace the clumsy `NPY_FAIL` / `*at = NULL` stuff internally to just use a regular return value, and write `PyArray_DescrConverter` as a thin wrapper.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Will add this to my pile of "things which didn't work today"... Splitting the first two commits to separate PRs, will revisit the third one once they're in.

File: numpy/core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -1651,71 +1565,186 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
...
+        return NPY_FAIL;
+    }
+}
+
+static int
+_convert_from_bytes(PyObject *obj, PyArray_Descr **at)
+{
+    int check_num = NPY_NOTYPE + 10;
+    int elsize = 0;
+    char endian = '=';
```
Comment (seberg) [inline numpy/core/src/multiarray/descriptor.c]: Dunno if its actually not worse, but `elsize` does not seem to need initialization (I think) and `endian` could also be a `default:` clause (editor always complains about missing default, but not sure I agree with it.

File: numpy/core/src/multiarray/descriptor.c [hunk 2]
```diff
@@ -1651,71 +1565,186 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
...
-                tmp = PyUnicode_FromEncodedObject(obj, "ascii", "strict");
-                if (tmp == NULL) {
-                    goto fail;
-                }
-                item = PyDict_GetItem(typeDict, tmp);
-                Py_DECREF(tmp);
-            }
-            else {
-                item = PyDict_GetItem(typeDict, obj);
+        if (PyBytes_Check(obj)) {
```
Comment (seberg) [inline numpy/core/src/multiarray/descriptor.c]: This check can be removed as well I think? Won't gh-15261 create a merge conflict here? (although I suppose its not a huge one).

**Review by seberg — APPROVED** [review]
LGTM, nice moving around much easier to read broken apart. The last if would be nice to remove, since you removed one further down? The other things I do not care about.

File: numpy/core/src/multiarray/descriptor.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: This is cleaned up in a future commit.

File: numpy/core/src/multiarray/descriptor.c (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: gh-15261 is on hold until I fix all this

File: numpy/core/src/multiarray/descriptor.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: Wait, no it isn't - pushed the change as an update. I've left them with initialization, but put the initialization right next to the use.

File: numpy/core/src/multiarray/descriptor.c (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: Good catch, check removed as part of the second commit in this patch. Also some other useless code elided.

**Review by seberg — APPROVED** [review]
Great, thanks. I think one of the declaration fixups came in the wrong commit, but I don't mind. I will merge this as soon as the tests all pass I think, since it will create merge conflicts with the Py3K cleanup, but it has to happen one way or the other...

**Comment by eric-wieser** [comment]
> I think one of the declaration fixups came in the wrong commit, 

Which one?

**Comment by seberg** [comment]
https://github.com/numpy/numpy/pull/15265/commits/6d2faa6f66e020379e3b80bd5678db4354aa76e9 commit, the removal of the duplicate `type` and `len` is in the next commit, but honestly I do not care, its a pretty history in either case, and only a compiler warning.

**Comment by eric-wieser** [comment]
> and only a compiler warning.

Nope, that's actually a bug in that intermediate commit. Happy to fix if you're happy to wait for CI again

**Comment by seberg** [comment]
OK, I don't mind waiting, but really a nitpick IMO.

**Comment by eric-wieser** [comment]
Proof the commits are the same: https://github.com/numpy/numpy/compare/5da2c01a7ae3b9fc3701801ab16d234e5425436c..ac44095f8f8aca2574091bb32d6328d0b6bbf29a

**Comment by eric-wieser** [comment]
CI passes, and this is an identical revision to the approved one, merging...

**Pull request merged**